pkg/gadgets/trace/dns/tracer/tracer.go

```diff
@@ -97,265 +67,147 @@ func NewTracer() (*Tracer, error) {
+	// The sample received is a concatenation of the dnsEventT structure and the packet bytes.
+	bpfEvent := (*dnsEventT)(unsafe.Pointer(&rawSample[0]))
+	packetBytes := rawSample[unsafe.Sizeof(*bpfEvent):]
+
+	// The DNS layer starts after the ethernet, IP and UDP headers.
+	// This is hardcoded for now as the datapath assumes the DNS packet is UDP over IPv4
+	// without any options. Next iterations will fix this.
+	// Ethernet + IP + UDP
+	const dnsOff = 14 + 20 + 8
+	dnsLayer := layers.DNS{}
+	err := dnsLayer.DecodeFromBytes(packetBytes[dnsOff:], gopacket.NilDecodeFeedback)
+	if err != nil {
+		return nil, fmt.Errorf("decoding dns layer: %w", err)
+	}
+
+	ipversion := gadgets.IPVerFromAF(bpfEvent.Af)
+
+	event := &types.Event{
 		Event: eventtypes.Event{
-			Type: eventtypes.NORMAL,
+			Type:      eventtypes.NORMAL,
+			Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
 		},
+		WithNetNsID:   eventtypes.WithNetNsID{NetNsID: netns},
+		WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MountNsId},
 		Pid:           bpfEvent.Pid,
 		Tid:           bpfEvent.Tid,
 		Uid:           bpfEvent.Uid,
 		Gid:           bpfEvent.Gid,
-		WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MountNsId},
-		WithNetNsID:   eventtypes.WithNetNsID{NetNsID: netns},
 		Comm:          gadgets.FromCString(bpfEvent.Task[:]),
-	}
-	event.Event.Timestamp = gadgets.WallTimeFromBootTime(bpfEvent.Timestamp)
+		PktType:       pktTypeToString(bpfEvent.PktType),
 
-	event.ID = fmt.Sprintf("%.4x", bpfEvent.Id)
-	ipversion := gadgets.IPVerFromAF(bpfEvent.Af)
-	event.SrcIP = gadgets.IPStringFromBytes(bpfEvent.SaddrV6, ipversion)
-	event.DstIP = gadgets.IPStringFromBytes(bpfEvent.DaddrV6, ipversion)
+		SrcIP:    gadgets.IPStringFromBytes(bpfEvent.SaddrV6, ipversion),
+		DstIP:    gadgets.IPStringFromBytes(bpfEvent.DaddrV6, ipversion),
+		SrcPort:  bpfEvent.Sport,
+		DstPort:  bpfEvent.Dport,
+		Protocol: gadgets.ProtoString(int(bpfEvent.Proto)),
+
+		ID:         fmt.Sprintf("%.4x", dnsLayer.ID),
+		NumAnswers: int(dnsLayer.ANCount),
+	}
 
-	if bpfEvent.Qr == 1 {
+	if dnsLayer.QR {
 		event.Qr = types.DNSPktTypeResponse
 		event.Nameserver = event.SrcIP
+
+		event.Latency = time.Duration(bpfEvent.LatencyNs)
+		event.Rcode = dnsLayer.ResponseCode.String()
 	} else {
 		event.Qr = types.DNSPktTypeQuery
 		event.Nameserver = event.DstIP
 	}
 
-	event.SrcPort = bpfEvent.Sport
-	event.DstPort = bpfEvent.Dport
-	event.Protocol = gadgets.ProtoString(int(bpfEvent.Proto))
-
-	// Convert name into a string with dots
-	event.DNSName = parseLabelSequence(bpfEvent.Name[:])
-
-	// Parse the packet type
-	event.PktType = "UNKNOWN"
-	pktTypeUint := uint(bpfEvent.PktType)
-	if pktTypeUint < uint(len(pktTypeNames)) {
-		event.PktType = pktTypeNames[pktTypeUint]
-	}
-
-	qTypeUint := uint(bpfEvent.Qtype)
-	var ok bool
-	event.QType, ok = qTypeNames[qTypeUint]
-	if !ok {
-		event.QType = "UNASSIGNED"
+	if len(dnsLayer.Questions) > 0 {
+		question := dnsLayer.Questions[0]
```

Comment:
IMO, it doesn't make sense to use gopacket parsing for some things and not for others. I agree gopacket is indeed a good source of trust, if we consider there is something missing, we should open a PR on gopacket instead of use a custom parsing. In this case in specific, honestly, I think people will mainly use the DNS gadget for tracing IPv4 and IPv6 request, and those are already included in the gopacket parsing. Said that, I suggest using the gopacket parsing also here.